third_party/xla_client/metrics_analysis.cc

```diff
@@ -0,0 +1,204 @@
+#include "tensorflow/compiler/xla/xla_client/tf_logging.h"
+#include "tensorflow/compiler/xla/xla_client/types.h"
+
+namespace xla {
+namespace metrics {
+
+namespace {
+
+float DivToFloat(std::ldiv_t val, std::int64_t denominator) {
+  return val.quot + (float) val.rem / denominator;
+}
+
+class MetricFrequency : public Analyzer {
+ public:
+  MetricFrequency(std::string metric_name, float frequency_threshold,
+                  long warmup_steps = 0)
+      : metric_name_(metric_name),
+        frequency_threshold_(frequency_threshold),
+        warmup_steps_(warmup_steps) {}
+
+  Analysis Run() override {
+    CounterData* step = GetCounter("MarkStep");
+    MetricData* metric = GetMetric(metric_name_);
+    if (!step || !metric) {
+      return {Analysis::Symptom::kNormal};
+    }
+    size_t metric_count = metric->TotalSamples();
+    xla::int64 step_count = step->Value();
+    if (step_count <= warmup_steps_) {
+      return {Analysis::Symptom::kNormal};
+    }
+    auto res = std::div(metric_count, step_count);
+    if (DivToFloat(res, step_count) > frequency_threshold_) {
+      return {
+          Analysis::Symptom::kMetricTooFrequent,
+          absl::StrFormat("%s too frequent: %zu counts during %zu steps",
+                          metric_name_, metric_count, step_count),
+      };
+    }
+    return {Analysis::Symptom::kNormal};
+  }
+
+ private:
+  std::string metric_name_;
+  float frequency_threshold_;
+  long warmup_steps_;
+};
+
+class MetricTime : public Analyzer {
+ public:
+  MetricTime(std::string metric_name, long threshdold_nsec)
+      : metric_name_(metric_name), threshold_nsec_(threshdold_nsec) {}
+
+  Analysis Run() override {
+    double max_metric_time = 0;
+    MetricData* metric = GetMetric(metric_name_);
+    if (!metric) {
+      return {Analysis::Symptom::kNormal};
+    }
+    // No need for accumulator and we want all recent samples.
+    for (const Sample& sample : metric->Samples(nullptr, nullptr)) {
+      max_metric_time = std::max(sample.value, max_metric_time);
+    }
+    if (max_metric_time > threshold_nsec_) {
+      return {
+          Analysis::Symptom::kMetricTooSlow,
+          absl::StrFormat("%s too slow: longest instance took %f seconds. "
+                          "Please open an GitHub issue with the graph dump for "
+                          "our team to optimize.",
+                          metric_name_, max_metric_time / 1e9),
+      };
+    }
+    return {Analysis::Symptom::kNormal};
+  }
+
+ private:
+  std::string metric_name_;
+  long threshold_nsec_;
+};
+
+class XrtMetricFrequency : public Analyzer {
+ public:
+  XrtMetricFrequency(std::map<std::string, float> metric_name_thresholds,
+                     int run_every_n)
+      : metric_name_thresholds_(std::move(metric_name_thresholds)),
+        run_every_n_(run_every_n),
+        counter_(0) {}
+
+  Analysis Run() override {
+    // XRT GetMetrics call is relatively expensive.
+    if (++counter_ % run_every_n_ != 0) {
```

Comment:
Maybe check `run_every_n_  !=0` here or check `run_every_n  != 0` in init time. We need to handle the edge case here or this function will never be called when `run_every_n_ == 0`

Reply:
Response in next comment.